File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    # 바뀌는 것을 value로 둠 (순위)
    rank = dict()
    for i in range(len(players)):
        rank[players[i]] = i
    
    for player in callings:
        nowIndex = rank.get(player)    
        players[nowIndex], players[nowIndex - 1] = players[nowIndex - 1], players[nowIndex]
        rank[players[nowIndex]] = nowIndex
        rank[players[nowIndex - 1]] = nowIndex - 1
    
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {   
        
        // 맵 형태로 저장 (선수이름 : 순위)
        HashMap<String, Integer> playerMap = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            playerMap.put(players[i], i);
        }
        
        for (int i = 0; i < callings.length; i++) {
            String playerName = callings[i];
            int playerRank = playerMap.get(playerName);
            
            // 앞 사람과 자리 교환
            String temp = players[playerRank - 1]; 
            players[playerRank - 1] = playerName; 
            players[playerRank] = temp;
            
            // 순위 업데이트
            playerMap.put(players[playerRank], playerRank);
            playerMap.put(players[playerRank - 1], playerRank - 1);
           
            // System.out.println(playerRank);
        }
        
        return players;
    }
}

// 이름이 불릴 때 앞 인덱스의 플레이어와 자리를 바꾸게 됨
// List, indexOf로 풀이 -> 시간 초과 -> o(1)로 찾을 수 있는 map으로 변환 
